### PR TITLE
feat(node:dns): implement `dns.lookupService`

### PR DESCRIPTION
**Pull request by Hanaasagi**
### What does this PR do?

Close: #4347 

Implement `dns.lookupService`

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [ ] Code changes

### How did you verify your code works?

I wrote automated tests

- [x] I ran `make js` and committed the transpiled changes
- [x] I or my editor ran Prettier on the changed files (or I ran `bun fmt`)
- [x] I included a test for the new code, or an existing test covers it
- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I or my editor ran `zig fmt` on the changed files
- [x] I included a test for the new code, or an existing test covers it
- [ ] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed

### CODE REVIEW

File: src/bun.js/api/bun/dns_resolver.zig [hunk 1]
```diff
@@ -2376,6 +2583,94 @@ pub const DNSResolver = struct {
...
+
+        const addr_s = addr_str.getZigString(globalThis).slice();
+        const port: u16 = if (port_value.isNumber()) blk: {
+            break :blk port_value.to(u16);
+        } else {
+            globalThis.throwInvalidArgumentType("lookupService", "port", "invalid port");
+            return .zero;
+        };
+
+        const addr = std.net.Address.parseIp(addr_s, port) catch {
```
Comment (Jarred-Sumner) [inline src/bun.js/api/bun/dns_resolver.zig]: Is there a c-ares function we can use for parsing ip addresses? I am not confident in the quality, reliability or security of `std.net.Address.parse*`

**Comment by Jarred-Sumner** [comment]
Thank you for this. It looks good. The only thing I'm concerned about is usage of `std.net.Address.parse*`. We should remove it from our codebase entirely. The other usages are in `isIPAddress` and `isIPV6Address`. If there aren't c-ares functions for this, there are very likely libc functions we can use which have been tested & used in production

File: src/bun.js/api/bun/dns_resolver.zig (hunk 1 above)
Comment (Hanaasagi) [reply src/bun.js/api/bun/dns_resolver.zig]: Fixed. Added a `getSockaddr` function to get the `sockaddr` struct, which uses `ares_inet_pton` to parse the address.

**Review by Jarred-Sumner — APPROVED** [review]
LGTM

File: test/js/node/dns/node-dns.test.js [hunk 2]
```diff
@@ -327,4 +327,59 @@ describe("test invalid arguments", () => {
...
+        expect(service).toStrictEqual(expected[1]);
+        done();
+      } catch (err) {
+        done(err);
+      }
+    });
+  });
+
+  it("lookupService(255.255.255.255, 443)", done => {
+    dns.lookupService("255.255.255.255", 443, (err, hostname, service) => {
```
Comment (Hanaasagi) [inline test/js/node/dns/node-dns.test.js:362]: on the darwin platform, the result is different, returning `[broadcasthost, https]`. I tested it with nodejs 18, and it behaves the same way.

https://github.com/Hanaasagi/darwin-ci/actions/runs/6233132116/job/16917692340#step:4:1

**Comment by Jarred-Sumner** [comment]
Thank you

**Pull request merged**